gflownet/envs/ctorus.py

```diff
@@ -143,54 +173,106 @@ def get_parents(
+        # Initialize actions tensor with EOS actions (inf) since these will be the
+        # actions for several special cases in both forward and backward actions.
+        actions_tensor = torch.full(
+            (n_states, self.n_dim), torch.inf, dtype=self.float, device=device
+        )
         # Sample angle increments
-        angles = torch.zeros(n_states, self.n_dim).to(device)
-        logprobs = torch.zeros(n_states, self.n_dim).to(device)
-        if torch.any(mask_states_sample):
+        if torch.any(do_sample):
             if sampling_method == "uniform":
                 distr_angles = Uniform(
                     torch.zeros(len(ns_range_noeos)),
                     2 * torch.pi * torch.ones(len(ns_range_noeos)),
                 )
             elif sampling_method == "policy":
-                mix_logits = policy_outputs[mask_states_sample, 0::3].reshape(
+                mix_logits = policy_outputs[do_sample, 0::3].reshape(
                     -1, self.n_dim, self.n_comp
                 )
                 mix = Categorical(logits=mix_logits)
-                locations = policy_outputs[mask_states_sample, 1::3].reshape(
+                locations = policy_outputs[do_sample, 1::3].reshape(
                     -1, self.n_dim, self.n_comp
                 )
-                concentrations = policy_outputs[mask_states_sample, 2::3].reshape(
+                concentrations = policy_outputs[do_sample, 2::3].reshape(
                     -1, self.n_dim, self.n_comp
                 )
                 vonmises = VonMises(
                     locations,
                     torch.exp(concentrations) + self.vonmises_min_concentration,
                 )
                 distr_angles = MixtureSameFamily(mix, vonmises)
-            angles[mask_states_sample] = distr_angles.sample()
-            logprobs[mask_states_sample] = distr_angles.log_prob(
-                angles[mask_states_sample]
-            )
+            angles_sampled = distr_angles.sample()
+            actions_tensor[do_sample] = angles_sampled
+            logprobs[do_sample] = distr_angles.log_prob(angles_sampled)
         logprobs = torch.sum(logprobs, axis=1)
-        # Build actions
-        actions_tensor = torch.inf * torch.ones(
-            angles.shape, dtype=self.float, device=device
-        )
-        actions_tensor[mask_states_sample, :] = angles[mask_states_sample]
+        # Catch special case for backwards return-to-source actions
+        if is_backward:
+            do_return_to_source = mask[:, 0]
+            if torch.any(do_return_to_source):
+                source_angles = tfloat(
+                    self.source[: self.n_dim], float_type=self.float, device=self.device
+                )
+                states_from_angles = tfloat(
+                    states_from, float_type=self.float, device=self.device
+                )[do_return_to_source, : self.n_dim]
+                actions_return_to_source = states_from_angles - source_angles
```

Comment:
Just to clarify, because there are multiple possible actions to reach the source (due to `% (2 * np.pi)`), the fact that we specify a particular one here shouldn't be a problem, right? (I don't think it should, but I recognize that's sort of an edge case, so just wondered if you thought about it.)

Reply:
Ok, false alarm.

Reply:
It's true that "there are multiple possible actions to reach the source". However, they all are the same action in practice because the VonMises distribution is defined in (-π, π).

Let's consider 1D for simplicity. Say we are in the the last state before the source going backwards, with angle θ, so state is [θ, 1]. There are infinite actions that would take us back to the source, but all have the same likelihood given a VonMises distribution defined by the parameters of the model, so it doesn't matter which one we choose. Let's consider the two actions with |Δθ| <= 2π:

- (-θ, )
- (2π - θ, )

Both have the same likelihood given a distribution. To check this empirically:

```
# VonMises distribution with mean at 90 degrees (0.5π) and concentration 1.
distr = VonMises(torch.tensor([0.5 * torch.pi]), torch.tensor([1.0]))
# θ = 0.5π
theta = torch.tensor(0.5 * torch.pi)
theta_inv = torch.tensor(2 * torch.pi) - theta
torch.isclose(distr.log_prob(-theta), distr.log_prob(theta_inv)) ->  tensor([True])
# θ = 0.12345
theta = torch.tensor(0.12345)
theta_inv = torch.tensor(2 * torch.pi) - theta
torch.isclose(distr.log_prob(-theta), distr.log_prob(theta_inv)) ->  tensor([True])
```

Does that make sense? Any remaining concerns?

Reply:
And actually, as discussed in the continuous GFlowNet paper, what we need to model is distribution over state transitions, from s_t to s_t+1 or vice versa, rather than over actions.